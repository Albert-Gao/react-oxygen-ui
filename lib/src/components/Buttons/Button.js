"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const styles_1 = require("../../styles");
const utils_1 = require("../../utils/utils");
const BaseBox_1 = require("../Box/BaseBox");
const shared_1 = require("./shared");
exports.Button = ({ noDefaultHoverCSS, children, css, disabled, takeRef, ...rest }) => {
    const cssToAdd = `
    ${shared_1.getBaseButtonCSS(noDefaultHoverCSS || disabled)};
    ${shared_1.baseNonFlatButtonCSS};
    ${styles_1.getSafeCss(css)};
  `;
    return (<BaseBox_1.BaseBox as="button" border={0} borderColor="transparent" borderRadius={0} fontSize={styles_1.DEFAULT_FONT_SIZE.bodyHeading} px={styles_1.DEFAULT_SPACE.small} minHeight="44px" maxHeight="44px" bg="primary" color="white" css={cssToAdd} {...utils_1.getRefProp(takeRef)} {...rest}>
      {children}
    </BaseBox_1.BaseBox>);
};
