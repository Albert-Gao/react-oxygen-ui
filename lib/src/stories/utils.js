"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const addon_actions_1 = require("@storybook/addon-actions");
const addon_knobs_1 = require("@storybook/addon-knobs");
const styles_1 = require("../styles");
function arrayToObject(data, keyToMap) {
    if (!data || data.length === 0) {
        return {};
    }
    const result = {};
    data.forEach((item, index) => {
        const i = keyToMap ? keyToMap[index] : index;
        result[i] = item;
    });
    return result;
}
exports.arrayToObject = arrayToObject;
// We wrap the action to prevent been display in addon-jsx
function wrapAction(actionName) {
    const wrapped = addon_actions_1.action(actionName);
    wrapped.toString = () => `action('${actionName}')`;
    return wrapped;
}
exports.wrapAction = wrapAction;
const keyOfColors = Object.keys(styles_1.defaultTheme.colors);
exports.Knobs = {
    spacesKnob: (propName, defaultValue = 2) => addon_knobs_1.select(propName, styles_1.DEFAULT_SPACE, defaultValue),
    colorKnob: (propName, defaultValue = 'primary') => addon_knobs_1.select(propName, arrayToObject(keyOfColors, keyOfColors), defaultValue),
    fontWeightKnob: (defaultValue = 'regular') => addon_knobs_1.select('fontWeight', {
        thin: 'thin',
        light: 'light',
        regular: 'regular',
        medium: 'medium',
        bold: 'bold',
        black: 'black',
    }, defaultValue),
    fontSizeKnob: (propName = 'fontSize', defaultValue = 1) => addon_knobs_1.select(propName, styles_1.DEFAULT_FONT_SIZE, defaultValue),
    sizeKnob: (propName, defaultValue = 'medium') => addon_knobs_1.select(propName, styles_1.DEFAULT_SIZE, styles_1.DEFAULT_SIZE[defaultValue]),
};
