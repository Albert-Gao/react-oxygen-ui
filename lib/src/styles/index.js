"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var fontSpecs_1 = require("./theme/constants/fontSpecs");
exports.DEFAULT_FONT_SIZE = fontSpecs_1.DEFAULT_FONT_SIZE;
exports.fontSpecs = fontSpecs_1.fontSpecs;
var spaces_1 = require("./theme/constants/spaces");
exports.DEFAULT_SPACE = spaces_1.DEFAULT_SPACE;
var sizes_1 = require("./theme/constants/sizes");
exports.DEFAULT_SIZE = sizes_1.DEFAULT_SIZE;
var colors_1 = require("./theme/constants/colors");
exports.DEFAULT_COLORS = colors_1.DEFAULT_COLORS;
var theme_1 = require("./theme/theme");
exports.defaultTheme = theme_1.defaultTheme;
__export(require("./theme/theme.utils"));
var styled_1 = require("./styled");
exports.styled = styled_1.styled;
exports.css = styled_1.css;
exports.createGlobalStyle = styled_1.createGlobalStyle;
exports.keyframes = styled_1.keyframes;
exports.ThemeProvider = styled_1.ThemeProvider;
var utils_1 = require("./utils");
exports.getTransitionValue = utils_1.getTransitionValue;
exports.getSafeValue = utils_1.getSafeValue;
exports.getSafeCss = utils_1.getSafeCss;
