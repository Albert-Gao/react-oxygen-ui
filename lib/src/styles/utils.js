"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const styled_system_1 = require("styled-system");
const utils_1 = require("../utils/utils");
exports.getMediaQuery = (screenSize, css) => `${styled_system_1.createMediaQuery(screenSize)} {
    ${css};
  }`;
exports.getTransitionValue = (propertyNames = 'transform') => `${propertyNames} 0.3s cubic-bezier(0.53, 0.01, 0.36, 1.63);`;
function getSafeValue(target, defaultValue) {
    return utils_1.is.notExist(target) ? defaultValue : target;
}
exports.getSafeValue = getSafeValue;
function getSafeCss(target) {
    return getSafeValue(target, '');
}
exports.getSafeCss = getSafeCss;
