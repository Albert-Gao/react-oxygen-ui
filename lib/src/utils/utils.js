"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.is = {
    function: (v) => typeof v === 'function',
    number: (v) => typeof v === 'number',
    notExist: (v) => v === undefined || v === null,
};
exports.getRefProp = (takeRef) => takeRef
    ? {
        ref: takeRef,
    }
    : {};
