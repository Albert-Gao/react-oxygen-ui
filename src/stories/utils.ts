import { action } from '@storybook/addon-actions';
import { select } from '@storybook/addon-knobs';
import { Responsive } from '../components';
import {
  ColorType,
  DEFAULT_FONT_SIZE,
  DEFAULT_SIZE,
  DEFAULT_SPACE,
  defaultTheme,
  FontWeightType,
} from '../styles';

export function arrayToObject<T>(
  data: T[],
  keyToMap?: string[],
): { [id: string]: T } {
  if (!data || data.length === 0) {
    return {};
  }

  const result: { [id: string]: T } = {};
  data.forEach((item, index) => {
    const i = keyToMap ? keyToMap[index] : index;
    result[i] = item;
  });

  return result;
}

// We wrap the action to prevent been display in addon-jsx
export function wrapAction(actionName: string) {
  const wrapped = action(actionName);
  wrapped.toString = () => `action('${actionName}')`;

  return wrapped;
}

const keyOfColors = Object.keys(defaultTheme.colors) as ColorType[];

export const Knobs = {
  spacesKnob: (propName: string, defaultValue: number = 2) =>
    select<number>(propName, DEFAULT_SPACE, defaultValue),

  colorKnob: (
    propName: string,
    defaultValue: ColorType = 'primary',
  ) =>
    select<string>(
      propName,
      arrayToObject<string>(
        keyOfColors as string[],
        keyOfColors as string[],
      ),
      defaultValue,
    ),

  fontWeightKnob: (defaultValue: FontWeightType = 'regular') =>
    select(
      'fontWeight',
      {
        thin: 'thin',
        light: 'light',
        regular: 'regular',
        medium: 'medium',
        bold: 'bold',
        black: 'black',
      },
      defaultValue,
    ),

  fontSizeKnob: (
    propName: string = 'fontSize',
    defaultValue: Responsive = 1,
    // @ts-ignore
  ) => select(propName, DEFAULT_FONT_SIZE, defaultValue),

  sizeKnob: (
    propName: string,
    defaultValue: keyof typeof DEFAULT_SIZE = 'medium',
  ) => select(propName, DEFAULT_SIZE, DEFAULT_SIZE[defaultValue]),
};
