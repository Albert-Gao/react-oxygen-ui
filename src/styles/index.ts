import { Responsive } from '../components';
import { defaultTheme } from './theme/theme';


export type ThemeType = typeof defaultTheme;
export type SpaceType = Responsive;
export type FontSizeType = Responsive;
export type GradientType = keyof ThemeType['gradients'];
export type ColorType = string | string[];
export type FontWeightType = keyof ThemeType['fontWeights'];
export type BoxShadowType = number | number[];

export { DEFAULT_FONT_SIZE, fontSpecs } from './theme/constants/fontSpecs';
export { DEFAULT_SPACE } from './theme/constants/spaces';
export { DEFAULT_SIZE } from './theme/constants/sizes';
export { DEFAULT_COLORS } from './theme/constants/colors';
export { defaultTheme } from './theme/theme';
export * from './theme/theme.utils';

export {
  styled,
  css,
  createGlobalStyle,
  keyframes,
  ThemeProvider,
} from './styled';

export {
  getTransitionValue,
  getSafeValue,
  getSafeCss,
} from './utils';
