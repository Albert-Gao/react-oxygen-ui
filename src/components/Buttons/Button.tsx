import * as React from 'react';
import { getSafeCss } from '../../styles/utils';
import { getRefProp } from '../../utils/utils';
import { BaseBox } from '../Box/BaseBox';
import { IBaseBoxWithTakeRef } from '../Box/BaseBox.type';
import { baseButtonCSS } from './shared';
import { css as cssFunc } from 'styled-components';

export interface IButton extends IBaseBoxWithTakeRef {
  disabled?: boolean;
}

export const Button: React.FC<IButton> = ({
  children,
  css,
  disabled,
  takeRef,
  ...rest
}) => {
  const cssToAdd = cssFunc`
    ${baseButtonCSS};
    text-transform: uppercase;
    overflow: hidden;
    ${getSafeCss(css)};
  `;

  return (
    <BaseBox as="button" css={cssToAdd} {...getRefProp(takeRef)} {...rest}>
      {children}
    </BaseBox>
  );
};
