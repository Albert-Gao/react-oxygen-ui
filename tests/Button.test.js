import 'jest-styled-components';
import 'jest-dom/extend-expect';
import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { Button } from '../src/components/Buttons/Button';
import { defaultTheme } from '../src/styles/IThemeType';
import { TestThemeProvider as ThemeProvider } from './utils';
import { css as cssFunc } from 'styled-components';

afterEach(cleanup);

describe('Test <Button />: ', () => {
  test('fontSize should be added', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Button fontSize="23px" />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();

    // the new css is there
    expect(firstChild).toHaveStyleRule('font-size', '23px');
  });
});

describe('Test <Button css={}/>: ', () => {
  test('css prop should be added', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Button css={'background-color: black;'} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
  });

  test('css prop (object) should be handled', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Button css={{ backgroundColor: 'black', paddingLeft: '1023px' }} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
    expect(firstChild).toHaveStyleRule('padding-left', '1023px');
  });

  test('css prop (css string) should be handled', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Button
          css={cssFunc`
          background-color: black;
          padding-left: 1023px;
          `}
        />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
    expect(firstChild).toHaveStyleRule('padding-left', '1023px');
  });
});
