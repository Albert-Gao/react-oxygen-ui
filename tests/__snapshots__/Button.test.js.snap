// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Button />:  fontSize should be added 1`] = `
.c0 {
  box-sizing: border-box;
  font-size: 23px;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
}

<button
  class="c0"
  font-size="23px"
/>
`;

exports[`Test <Button css={}/>:  css prop (css string) should be handled 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
  background-color: black;
  padding-left: 1023px;
}

<button
  class="c0"
/>
`;

exports[`Test <Button css={}/>:  css prop (object) should be handled 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
  background-color: black;
  padding-left: 1023px;
}

<button
  class="c0"
/>
`;

exports[`Test <Button css={}/>:  css prop should be added 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  display: inline-block;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: 0;
  text-align: center;
  line-height: inherit;
  white-space: nowrap;
  cursor: pointer;
  text-transform: uppercase;
  overflow: hidden;
  background-color: black;
}

<button
  class="c0"
/>
`;
